keygen/src/keygen.rs

```diff
@@ -310,78 +374,167 @@ fn main() -> Result<(), Box<dyn error::Error>> {
+                values_t_or_exit!(matches, "starts_with", String)
                     .into_iter()
                     .map(|s| if ignore_case { s.to_lowercase() } else { s })
                     .collect()
             } else {
                 HashSet::new()
             };
-
-            let starts_with = if matches.is_present("starts_with") {
-                values_t_or_exit!(matches, "starts_with", String)
+            let ends_with_args = if matches.is_present("ends_with") {
+                values_t_or_exit!(matches, "ends_with", String)
+                    .into_iter()
+                    .map(|s| if ignore_case { s.to_lowercase() } else { s })
+                    .collect()
+            } else {
+                HashSet::new()
+            };
+            let starts_and_ends_with_args = if matches.is_present("starts_and_ends_with") {
+                values_t_or_exit!(matches, "starts_and_ends_with", String)
                     .into_iter()
                     .map(|s| if ignore_case { s.to_lowercase() } else { s })
                     .collect()
             } else {
                 HashSet::new()
             };
 
-            if includes.is_empty() && starts_with.is_empty() {
+            if starts_with_args.is_empty() && ends_with_args.is_empty() && starts_and_ends_with_args.is_empty() {
                 eprintln!(
-                    "Error: No keypair search criteria provided (--includes or --starts-with)"
+                    "Error: No keypair search criteria provided (--starts-with or --ends-with or --starts-and-ends-with)"
                 );
                 exit(1);
             }
 
+            for sw in &starts_with_args {
+                let args: Vec<&str> = sw.split(':').collect(); 
+                grind_matches.push(Match{
+                    starts: args[0].to_lowercase(),
+                    ends: "".to_string(),
+                    count: args[1].parse::<u64>().unwrap(),
+                    found: AtomicU64::new(0)
+                });
+            }
+            for ew in &ends_with_args {
+                let args: Vec<&str> = ew.split(':').collect(); 
+                grind_matches.push(Match{
+                    starts: "".to_string(),
+                    ends: args[0].to_lowercase(),
+                    count: args[1].parse::<u64>().unwrap(),
+                    found: AtomicU64::new(0)
+                });
+            }
+            for swew in &starts_and_ends_with_args {
+                let args: Vec<&str> = swew.split(':').collect(); 
+                grind_matches.push(Match{
+                    starts: args[0].to_lowercase(),
+                    ends: args[1].to_lowercase(),
+                    count: args[2].parse::<u64>().unwrap(),
+                    found: AtomicU64::new(0)
+                });
+            }
+
+            let mut starting_messages = Vec::<String>::new();
+            for i in 0..grind_matches.len() {
```

Comment:
I think in this loop it would be fine to print the last message for all the grind_matches:
```
starting_messages.push(format!("{} {} that starts with '{}' and ends with '{}'",grind_matches[i].count, pk,grind_matches[i].starts,grind_matches[i].ends))
```
 We'd see the empty string if a --starts-with or --ends-with was provided, but it simplifies the code here a ton.

Reply:
updated